tsinfer/eval_util.py

```diff
@@ -756,104 +749,126 @@ def mean_sample_ancestry(ts, sample_sets, show_progress=False):
     Computes the mean sample ancestry for each node in the tree sequence with
     respect to the specified list of sets of samples, returning a 2D array with
     dimensions (len(sample_sets), ts.num_nodes). For a given element of this
-    array, A[k, u] is the average fraction of samples that descend from u
-    that are from samples_sets[k]. The average for each node is computed
-    by weighting the contribution along the span of u by the distance it
-    persists unchanged.
-    """
-    num_sample_sets = len(sample_sets)
-    S = np.zeros(ts.num_nodes)
-    A = np.zeros((num_sample_sets, ts.num_nodes))
-    span_start = np.zeros(ts.num_nodes) - 1
-    last_update = np.zeros(ts.num_nodes) - 1
-    tree_iters = zip(*[ts.trees(tracked_samples=samples) for samples in sample_sets])
-    trees = None
-    progress_iter = tqdm.tqdm(
-        zip(ts.edge_diffs(), ts.trees()),
-        total=ts.num_trees, disable=not show_progress)
-
-    for diffs, tree in progress_iter:
-        (left, right), edges_out, edges_in = diffs
-        assert tree.interval == (left, right)
+    array, A[k, u] is the average fraction of samples (within the entire set of
+    samples specified) descending from u that are from samples_sets[k]. The average
+    for each node is computed by weighting the contribution along the span of
+    u by the distance it persists unchanged.
+    """
```

Comment:
I thought you might be interested in this @petrelharp --- I've simplified this algorithm a lot and hopefully put it in a structure that will generalise to other stats. It's clearly going to be fast too - basically just a small constant on top of computing the sample counts.